### ISSUE TITLE
Invalid memory usage and crashes when importing broken fbx files

### ISSUE THREAD

**Issue by qarmin** [godotengine/godot#46530]
**Godot version:**
3.2.4.rc.custom_build. 8385a0d8a

**Issue description:**
```
=================================================================
==80950==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x62b002eac9ec at pc 0x7f91e38386aa bp 0x7ffc61190090 sp 0x7ffc6118f838
READ of size 16777229 at 0x62b002eac9ec thread T0
    #0 0x7f91e38386a9 in __interceptor_memcpy (/lib/x86_64-linux-gnu/libasan.so.6+0x3a6a9)
    #1 0x4312ffa in void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char const*>(char const*, char const*, std::forward_iterator_tag) /usr/include/c++/10/bits/basic_string.tcc:225
    #2 0x4296233 in FBXDocParser::Token::Token(char const*, char const*, FBXDocParser::TokenType, unsigned long) modules/fbx/fbx_parser/FBXBinaryTokenizer.cpp:133
    #3 0x429a335 in ReadScope modules/fbx/fbx_parser/FBXBinaryTokenizer.cpp:388
    #4 0x429ad36 in ReadScope modules/fbx/fbx_parser/FBXBinaryTokenizer.cpp:413
    #5 0x429ad36 in ReadScope modules/fbx/fbx_parser/FBXBinaryTokenizer.cpp:413
    #6 0x429ad36 in ReadScope modules/fbx/fbx_parser/FBXBinaryTokenizer.cpp:413
    #7 0x429ad36 in ReadScope modules/fbx/fbx_parser/FBXBinaryTokenizer.cpp:413
    #8 0x429bed5 in FBXDocParser::TokenizeBinary(std::vector<FBXDocParser::Token*, std::allocator<FBXDocParser::Token*> >&, char const*, unsigned long) modules/fbx/fbx_parser/FBXBinaryTokenizer.cpp:464
    #9 0x30606c1 in EditorSceneImporterFBX::import_scene(String const&, unsigned int, int, List<String, DefaultAllocator>*, Error*) modules/fbx/editor_scene_importer_fbx.cpp:127
    #10 0x9bb71e0 in ResourceImporterScene::import(String const&, String const&, Map<StringName, Variant, Comparator<StringName>, DefaultAllocator> const&, List<String, DefaultAllocator>*, List<String, DefaultAllocator>*, Variant*) editor/import/resource_importer_scene.cpp:1317
    #11 0x86494de in EditorFileSystem::_reimport_file(String const&) editor/editor_file_system.cpp:1788
    #12 0x8654ad6 in EditorFileSystem::reimport_files(Vector<String> const&) editor/editor_file_system.cpp:1986
    #13 0x861341a in EditorFileSystem::_update_scan_actions() editor/editor_file_system.cpp:589
    #14 0x862b885 in EditorFileSystem::_notification(int) editor/editor_file_system.cpp:1163
    #15 0x866504d in EditorFileSystem::_notificationv(int, bool) (/usr/bin/godots+0x866504d)
    #16 0x11538d4b in Object::notification(int, bool) core/object.cpp:929
    #17 0xc163a11 in SceneTree::_notify_group_pause(StringName const&, int) scene/main/scene_tree.cpp:992
    #18 0xc153f9f in SceneTree::idle(float) scene/main/scene_tree.cpp:529
    #19 0x18dcd08 in Main::iteration() main/main.cpp:2115
    #20 0x17c08fc in OS_X11::run() platform/x11/os_x11.cpp:3628
    #21 0x172cefb in main platform/x11/godot_x11.cpp:56
    #22 0x7f91e27350b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)
    #23 0x172cb1d in _start (/usr/bin/godots+0x172cb1d)

0x62b002eac9ec is located 0 bytes to the right of 26604-byte region [0x62b002ea6200,0x62b002eac9ec)
allocated by thread T0 here:
    #0 0x7f91e38ae517 in malloc (/lib/x86_64-linux-gnu/libasan.so.6+0xb0517)
    #1 0x11a7abdf in Memory::alloc_static(unsigned long, bool) core/os/memory.cpp:82
    #2 0x19b79ed in PoolVector<unsigned char>::resize(int) core/pool_vector.h:584
    #3 0x305f686 in EditorSceneImporterFBX::import_scene(String const&, unsigned int, int, List<String, DefaultAllocator>*, Error*) modules/fbx/editor_scene_importer_fbx.cpp:106
    #4 0x9bb71e0 in ResourceImporterScene::import(String const&, String const&, Map<StringName, Variant, Comparator<StringName>, DefaultAllocator> const&, List<String, DefaultAllocator>*, List<String, DefaultAllocator>*, Variant*) editor/import/resource_importer_scene.cpp:1317
    #5 0x86494de in EditorFileSystem::_reimport_file(String const&) editor/editor_file_system.cpp:1788
    #6 0x8654ad6 in EditorFileSystem::reimport_files(Vector<String> const&) editor/editor_file_system.cpp:1986
    #7 0x861341a in EditorFileSystem::_update_scan_actions() editor/editor_file_system.cpp:589
    #8 0x862b885 in EditorFileSystem::_notification(int) editor/editor_file_system.cpp:1163
    #9 0x866504d in EditorFileSystem::_notificationv(int, bool) (/usr/bin/godots+0x866504d)
    #10 0x11538d4b in Object::notification(int, bool) core/object.cpp:929
    #11 0xc163a11 in SceneTree::_notify_group_pause(StringName const&, int) scene/main/scene_tree.cpp:992
    #12 0xc153f9f in SceneTree::idle(float) scene/main/scene_tree.cpp:529
    #13 0x18dcd08 in Main::iteration() main/main.cpp:2115
    #14 0x17c08fc in OS_X11::run() platform/x11/os_x11.cpp:3628
    #15 0x172cefb in main platform/x11/godot_x11.cpp:56
    #16 0x7f91e27350b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)

SUMMARY: AddressSanitizer: heap-buffer-overflow (/lib/x86_64-linux-gnu/libasan.so.6+0x3a6a9) in __interceptor_memcpy
```
```
terminate called after throwing an instance of 'std::length_error'
  what():  basic_string::_M_create
Aborted (core dumped)
```

**Minimal reproduction project:**
3 Files which broke Godot - [Broken.zip](https://github.com/godotengine/godot/files/6058039/Broken.zip)
~1000 files in which a lot of files are broken and importing this models in some situations will cause crash/invalid memory usage - [a.zip](https://github.com/godotengine/godot/files/6058046/a.zip)

**Comment by RevoluPowered** [comment]
Yep I fully agree its time to get rid of this problem once and for all #47269

**Comment by RevoluPowered** [comment]
@qarmin need some help testing this out

#47880 

I fixed the 3 files mentioned here that crashed. Blender can't load them either, so we now have error checking in the parser.

Valgrind won't run on my macbook pro anymore, only ubsan works...

**Comment by akien-mga** [comment]
Fixed by #47880.
